Move state from workflow into context

kbrock opened the pull request:
References:
- Extracted from #67
- built upon (to remove codeclimate / failing build) #89

Goal
====

Move the runtime variables from `Workflow` to `Context`. So context will hold all the stateful data. No more methods in workflow that hold some context state.

Tangible goal:

- We can test the various `State` implementations without a `Workflow` instance.
- `State#run` will fully change the context and 

Additions
=======

- Added `NextState`, `Error`, and `Cause` to `Context`. Now errors are exposed.
- Added `context.execution["EndTime"]` and use that for determining `end?` (couldn't find the exact value for that variable)

Changes
=======

- Move hardcoded `State#status` methods to dynamic `Context#status`.
- Definition of `status "pending"` changed a little: it asks "Have I started" instead of "Is my state the starting state".
- Definition of `status "success"` changed a little: it asks "Have I **finished** without error" rather than looking at (`State#next` / `Context#next_state`).

Introduce Facade
=========

I introduced facade `quick_step`. This simplifies the caller in manageiq-provider-workflow's [workflow_instance.rb]:

```ruby
context = Workflow.quick_step(name, payload, context, creds)
update!(:context => context.to_h, :status => context.status, :output => context.output)
```

It is no secret from #67 , that I would like to pass the `context` into `step!` and feel that we could introduce a more sophisticated `step` method - which we can discuss elsewhere.

[workflow_instance.rb]: https://github.com/ManageIQ/manageiq-providers-workflows/blob/3a0900d3435875abd2d8dc9b2e29dcb57d374ecc/app/models/manageiq/providers/workflows/automation_manager/workflow_instance.rb#L71-L74

agrare commented on lib/floe/workflow/context.rb:
```diff
@@ -21,10 +21,46 @@ def execution
         @context["Execution"]
       end
 
+      def start?
+        execution.key?("StartTime")
+      end
+
+      def end?
+        execution.key?("EndTime")
+      end
```
Would `started?` and `running?` make more sense?  `context.running?` I think is a little more obvious than `context.end?`

kbrock replied:
I changed `started?` and `running?` - but left `end?` alone since It is part of the external interface.

Once we change `workflow_instance#run` to `context.running?`, then we can probably drop `context#end?`

agrare commented on lib/floe/workflow/context.rb:
```diff
@@ -21,10 +21,46 @@ def execution
+        elsif state["Error"]
+          "failure"
+        else
+          "success"
```
Hm if one step has finished (and thus has set execution["EndTime"] for _that_ state) will this report that the whole workflow is status "success" ?

Should we be clearing the execution Start and End times after the state has run?

agrare replied:
Ah nevermind execution["EndTime"] is only if there is no next state

agrare commented:
Overall I like this a lot, helped me with the async runner so I based my branch off of this one

kbrock commented:
update:
- rebased
- renamed `Context#start?`, `Context#end?` to `started?`, `ended?` and added `running?`
- removed `quick_run`
- added a few specs around `ended?`
